### DISCUSSION TITLE
Google Assistant SDK - Rspi 3 - "sounddevice.PortAudioError: Error querying device -1"

### DISCUSSION THREAD

**Discussion by kamortega** [Arkq/bluez-alsa#603] [Q&A]
Hi, 
Installed google assistant sdk on raspi3, the speaker is a home mini bluetooth, it is paired and connected to raspi, played from youtube and it works! even google says it's connected! 
 
However, when running command in terminal as (env) "googlesamples-assistant-pushtotalk --project-id --device-model-id"  I get the following: 

/home/pi/env/lib/python3.5/site-packages/google/auth/crypt/_cryptography_rsa.py:22: CryptographyDeprecationWarning: Python 3.5 support will be dropped in the next release of cryptography. Please upgrade your Python.
  import cryptography.exceptions
INFO:root:Connecting to embeddedassistant.googleapis.com
Traceback (most recent call last):
  File "/home/pi/env/bin/googlesamples-assistant-pushtotalk", line 8, in <module>
    sys.exit(main())
  File "/home/pi/env/lib/python3.5/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/pi/env/lib/python3.5/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/pi/env/lib/python3.5/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/pi/env/lib/python3.5/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/pi/env/lib/python3.5/site-packages/googlesamples/assistant/grpc/pushtotalk.py", line 351, in main
    flush_size=audio_flush_size
  File "/home/pi/env/lib/python3.5/site-packages/googlesamples/assistant/grpc/audio_helpers.py", line 190, in __init__
    blocksize=int(block_size/2),  # blocksize is in number of frames.
  File "/home/pi/env/lib/python3.5/site-packages/sounddevice.py", line 1345, in __init__
    **_remove_self(locals()))
  File "/home/pi/env/lib/python3.5/site-packages/sounddevice.py", line 762, in __init__
    samplerate)
  File "/home/pi/env/lib/python3.5/site-packages/sounddevice.py", line 2571, in _get_stream_parameters
    info = query_devices(device)
  File "/home/pi/env/lib/python3.5/site-packages/sounddevice.py", line 569, in query_devices
    raise PortAudioError('Error querying device {0}'.format(device))
sounddevice.PortAudioError: Error querying device -1

When using aplay -l in terminal: 
"aplay: device_list:270: no soundcards found..."

Also, running in terminal using "speaker-test -t wav",  the test runs but no sound is working"
" speaker-test 1.1.3
Playback device is default
Stream parameters are 48000Hz, S16_LE, 1 channels
WAV file(s)
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 9600 to 4194304
Period size range from 480 to 4096
Using max buffer size 4194304
Periods = 4
was set period_size = 4096
was set buffer_size = 4194304
 0 - Front Left
Time per period = 0.339021
 0 - Front Left
Time per period = 0.315553
 0 - Front Left
Time per period = 0.315577
*Keeps generating but no sound." 

Finally, going through sudo nano /home/pi/.asoundrc file, when connected to speaker is: 
pcm.!default {
        type plug
        slave.pcm {
                type bluealsa
                device "x:x:x:x:x:x"
                profile "a2dp"
        }
}
ctl.!default {
        type bluealsa
}

AND when going to "sudo nano /etc/asound.conf" generates another code, when connected to same speaker: 
pcm.!default {
type asym
capture.pcm "mic"
playback.pcm "speaker"
}
pcm.mic {
type plug
slave.pcm {
type bluealsa device "x:x:x:x:x:x"
profile "sco"

}
}
pcm.speaker {
type plug
slave.pcm {
type bluealsa device "x:x:x:x:x:x"
profile "sco"
}
}

 I tried copy/pasting code of  /etc/asound.conf in /home/pi/.asoundrc and run speaker-test -t wav, but : 
 "speaker-test 1.1.3

Playback device is default
Stream parameters are 48000Hz, S16_LE, 1 channels
WAV file(s)
ALSA lib bluealsa-pcm.c:680:(_snd_pcm_bluealsa_open) Couldn't get BlueALSA transport: No such device
Playback open error: -19,No such device" 

So, whats the deal??

**Comment by borine** [comment]
> ALSA lib bluealsa-pcm.c:680:(_snd_pcm_bluealsa_open) Couldn't get BlueALSA transport: No such device

That error message suggests you are using BlueALSA version 1.3.0 which was released in  July 2018. With the best will in the world, it is not possible for a small project like this to support releases that are more than 4 years old. If you are tied to that release by your distribution, then please seek help from that distribution, or else update your installation to the most recent release (v4.0.0) or the latest development sources from this repository.

With regard to speaker-test, even the latest BlueALSA sources have an issue there. As a workaround it is necessary to give an explicit buffer size on the command line, eg:
```
speaker-test -D bluealsa --buffer=500000
```
